### PR TITLE
Implement __array_function__ interface

### LINKED ISSUE #1322: RFC __array_function__ protocol
Continuing on the theme of #1247 CuPy developers may be interested in the following proposed Numpy Enhancement Proposal (NEP) to allow all Numpy functions (not just ufuncs) to work on other Numpy-like projects like CuPy

-  [NEP 0018: Dispatch Mechanism for NumPy’s high level API](http://www.numpy.org/neps/nep-0018-array-function-protocol.html)

This is being discussed on the Numpy mailing list here:

-  https://mail.python.org/pipermail/numpy-discussion/2018-June/078127.html

It would be useful to get the perspective of CuPy developers in this discussion.  

Feel free to close this issue immediately, I just wanted to inform the CuPy developers of this discussion.

### PR DESCRIPTION
**Pull request by okuta**
Close #1322 
Please see discussion ~numpy/numpy#11129~ numpy/numpy#12028.

### CODE REVIEW

File: cupy/core/core.pyx [hunk 1]
```diff
@@ -1718,6 +1718,14 @@ cdef class ndarray:
         else:
             return NotImplemented
 
+    def __array_function__(self, func, types, args, kwargs):
+        if not hasattr(cupy, func.__name__):
+            return NotImplemented
+        for t in types:
+            if not isinstance(t, ndarray):
```
Comment (hameerabbasi) [inline cupy/core/core.pyx]: Should be `if not all(t in HANDLED_TYPES for t in types):` where `HANDLED_TYPES` is something like `[np.ndarray, cupy_array]`.

Comment (okuta) [reply cupy/core/core.pyx]: Thanks, I fixed.

**Comment by okuta** [comment]
@hameerabbasi I want to try with NumPy. Do you have an implementation of NEP 18?

**Comment by shoyer** [comment]
We don't have any implementation of NEP 18 yet, beyond the [proof of concept implementation](https://nbviewer.jupyter.org/gist/shoyer/1f0a308a06cd96df20879a1ddb8f0006) referenced in the NEP (which doesn't actually implement the NumPy API).

**Comment by okuta** [comment]
Please read numpy/numpy#12028.

**Comment by shoyer** [comment]
The development branch of numpy now has support for `__array_function__`. It would be great if you could give this a try and report back any issues!

**Comment by okuta** [comment]
Thanks, I will try it.

**Comment by okuta** [comment]
This PR is working with NumPy master branch. @shoyer 
```
import numpy
import cupy

a = cupy.arange(10).reshape(2,5)
print(numpy.argmax(a))         #=> array(9)
print(numpy.argmax(a,axis=0))  #=> array([1, 1, 1, 1, 1])
```

**Comment by takagi** [comment]
LGTM! I suppose that it is reasonable to merge this feature soon as `__array_function__` will be experimentally supported since comming NumPy 1.16.0.

**Pull request merged**